### PR TITLE
rpc proxy

### PR DESCRIPTION
**Pull request by calvogenerico**
# What ❔

Proxy for rpc

## Why ❔

We need to provide a way to access blockchain data but with access control

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: packages/private-rpc/example-permissions.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      - "0x72D8dd6EE7ce73D545B229127E72c8AA013F4a9e"
+  - name: "admins"
+    members:
+      - "0xC8a6f2deD7FC49a5d53AC84A2AF5B1e5147Eb466"
+
+contracts:
+  - address: "0x0000000000000000000000000000000000008008"
+    methods:
+      - selector: "0x62f84b24"
+        read: "public"
```
Comment (aon) [inline packages/private-rpc/example-permissions.yaml:14]: Perhaps a better name for wildcard permissions could be `*`

File: packages/private-rpc/example-permissions.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        write: "public"
+  - address: "0xAe45cBE2d1E90358CbD216bC16f2C9267a4EA80a"
+    methods:
+      - signature: "function _checkOwner() internal view virtual" # if it's not a hex its interpreted as function signature
```
Comment (aon) [inline packages/private-rpc/example-permissions.yaml:18]: The name of the property already implies it's a signature and not a selector.

File: packages/private-rpc/example-permissions.yaml
Comment (aon) [inline packages/private-rpc/example-permissions.yaml:1]: Love the permission file, very clear and concise 🚀

File: packages/private-rpc/src/db/hex-row.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  toDriver(val) {
+    const parsed = hexSchema.parse(val);
+    const hex = parsed.slice(2);
+    // Avoid sending odd lengths. When length is odd Buffer.from ignores the first digit.
+    const prefix = hex.length % 2 === 0 ? '' : '0';
+    return Buffer.from(`${prefix}${hex}`, 'hex');
+  },
+  fromDriver(val) {
+    const hex = `0x${val.toString('hex')}`;
+    return hexSchema.parse(hex) as Hex;
```
Comment (aon) [inline packages/private-rpc/src/db/hex-row.ts:26]:
```suggestion
    return hexSchema.parse(hex);
```

File: packages/private-rpc/src/permissions/authorizeer.ts
Comment (aon) [inline packages/private-rpc/src/permissions/authorizeer.ts:1]: File name

File: packages/private-rpc/src/permissions/yaml-parser.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  AccessRule,
+  GroupRule,
+  Permission,
+  PublicRule,
+} from '@/permissions/access-rules';
+
+const methodSchema = z
+  .object({
+    selector: z.optional(hexSchema),
+    signature: z.optional(z.string()),
```
Comment (aon) [inline packages/private-rpc/src/permissions/yaml-parser.ts:16]: We can probably add another validation for the signature field as well using some viem util.

File: packages/private-rpc/src/permissions/yaml-parser.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    read: z.union([z.literal('public'), z.array(z.string())]),
+    write: z.union([z.literal('public'), z.array(z.string())]),
```
Comment (aon) [inline packages/private-rpc/src/permissions/yaml-parser.ts:18]: Why particularly separate "public", as it's still a string?

File: packages/private-rpc/src/rpc/rpc-method-handlers.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  max_fee_per_gas: hexSchema.optional(),
+  max_priority_fee_per_gas: z.number().optional(),
+  value: hexSchema.optional(),
+  data: hexSchema.optional(),
+  input: hexSchema.optional(),
+  nonce: hexSchema.optional(),
+  transaction_type: z.number().optional(),
+  access_list: z.any().optional(),
+  customData: z.any().optional(),
+});
```
Comment (aon) [inline packages/private-rpc/src/rpc/rpc-method-handlers.ts:30]: I believe it's not really necessary to parse all the request, just the values that we use. This way we: don't write so much code and prevent ourselves from modifying the data sent by the user

File: packages/private-rpc/src/rpc/delegate-call.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      body.error.message,
+      body.error.data,
+    );
+  }
+
+  if (body.result === undefined) {
+    throw new ExternalRpcError(-32000, 'Error', undefined);
+  }
+
+  return body?.result;
```
Comment (aon) [inline packages/private-rpc/src/rpc/delegate-call.ts:38]: Why do we need to interfere and get in the way of errors coming from the actual RPC? The client already knows how to handle them. In this case I would just pass them through as they are.

File: packages/private-rpc/src/permissions/yaml-parser.ts (hunk 5 above)
Comment (calvogenerico) [reply packages/private-rpc/src/permissions/yaml-parser.ts:18]: I made it like you can put `"public"` as an str __or__ `["addr1", "addr2"]` as a list of strings. That's why I used the union, to sepparate `string` from `Array<string>`

File: packages/private-rpc/src/rpc/rpc-method-handlers.ts (hunk 6 above)
Comment (calvogenerico) [reply packages/private-rpc/src/rpc/rpc-method-handlers.ts:30]: I believe that I tried and zod was returning only the specified arguments. Is there any way to make it preserve the original object?

File: packages/private-rpc/src/rpc/delegate-call.ts (hunk 7 above)
Comment (calvogenerico) [reply packages/private-rpc/src/rpc/delegate-call.ts:38]: That's true. Let me change this.

File: packages/private-rpc/src/permissions/yaml-parser.ts (hunk 5 above)
Comment (aon) [reply packages/private-rpc/src/permissions/yaml-parser.ts:18]: Makes total sense

File: packages/private-rpc/src/rpc/rpc-method-handlers.ts (hunk 6 above)
Comment (calvogenerico) [reply packages/private-rpc/src/rpc/rpc-method-handlers.ts:30]: Awesome.

**Review by aon — APPROVED** [review]

**Pull request merged**